Added links to blog cards and displays the blogposts. Need debubbing …

BigSpaces opened the pull request:
…and no responsive yet.

Would love to have a session if you have the time, and go over the updates. Some things, specially relating to how Phoenix 1.7 handles routes, are not too clear and I am quite sure I am not following the rulebook :)

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+        <p class="mb-6 text-2xl"><%= raw @blog.description %></p>
+        <p class="mb-6 text-l">By <%= raw @blog.author_name %>. Written on <%= Date.to_string(@blog.date) %> </p>
+        <img class="mb-4 rounded-2xl" src={"/images/posts/#{@blog.cover_image}"}/>
+        <%!-- {#if @blog.livebook_url}
```
Make sure to remove all commented code. I think this is a WIP so it's not too bad to have them in, but generally, try not to commit commented code.

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+
+  def mount(%{"id" => id}, _session, socket) do
+
+   blog = ElixirNewbie.Blog.get_post_by_id!(Blog.posts_and_drafts(), id) ## BigSpaces Sherlock Mode: Refactor or re-think this. Are we passing ALL of the posts?
```
Mount is called twice, so the Blog context is called twice. That's not a huge deal but could be a performance concern.

BigSpaces replied:
My concern is that I am not too sure I understand what is happening when `Blog.posts_and_drafts()` is passed to `ElixirNewbie.Blog.get_post_by_id!`

I assumed that `ElixirNewbie.Blog.get_post_by_id!` received an ID and it returned the blogpost, but when I insptected, it seems that all of the blogposts are being passed?

BrooklinJazz replied:
get_post_by_id takes in a list of posts and finds the post that matches the id. I can't remember why it's built this way instead of just receiving the list itself but there is (probably) a good reason for it. Maybe for testing purposes 🤔

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+  def mount(%{"id" => id}, _session, socket) do
+
+   blog = ElixirNewbie.Blog.get_post_by_id!(Blog.posts_and_drafts(), id) ## BigSpaces Sherlock Mode: Refactor or re-think this. Are we passing ALL of the posts?
+   IO.inspect(blog, label: "POSSSST")
```
Remove IO.inspect from PRs

BigSpaces replied:
Corrected

BrooklinJazz commented on lib/elixir_newbie_web/router.ex:
```diff
@@ -19,6 +19,7 @@ defmodule ElixirNewbieWeb.Router do
     # get "/", PageController, :home
     live "/", HomeLive, :home
     live "/blog", BlogLive, :blog
+    live "/blog/:id", BlogShow
```
While not strictly necessary, I think we should provide a live-action atom for the sake of consistency with the other pages.

That, or we should remove the live actions from the other routes.

Consistency for the sake of consistency is sometimes important because right now, the reader asks "why does this one not need the atom?" which may lead to wasted time on a question that has no answer.

BrooklinJazz replied:
Otherwise this is exactly how to handle routing, you mentioned you wanted feedback and my feedback is it looks good!

BigSpaces replied:
Yes! I did not manage to make it work with the atom and, after a bit of tinkering, I got it to a working state, even though it did not seem consistent with the other routes.

If I add an atom I get the following error and I have not yet managed to move past it:

`could not infer :as option because a live action was given and the LiveView does not have a "Live" suffix. Please pass :as explicitly or make sure your LiveView is named like "FooLive" or "FooLive.Index"`

My concerns with routing were in `blog_card.ex`: the hyperlink in the blog card is generated this way:

`<a href={static_url(ElixirNewbieWeb.Endpoint, "/blog/#{@blog.id}")}>`

Now I have changed it for 

`<.link navigate={"/blog/#{@blog.id}"}>`

I assume the latter is better?

BrooklinJazz reviewed (changes requested):
The overall direction looks great. Excellent work figuring this out Jon and very well put together.

I made some nitpicky comments, that I'm not sure you were ready for or not.
In the future I recommend adding either WIP or DRAFT in the title of your PR, or including some comments about the type of feedback you want and don't want. For example, if you know you have some IO inspects or commented code, and want me to ignore that then you can mention it in the PR description.

There are a couple of UI bugs I found

* Images don't have enough spacing in .markdown
* Some images at the top of the blog aren't full-width.
* H1 elements in the .markdown don't have enough spacing

BrooklinJazz commented:
Bug Reports:

Blog Images Not Full Width:
![image](https://github.com/BrooklinJazz/elixir_newbie_website_v2/assets/14877564/2eecf3be-6e12-4c10-b103-39f954254e11)

H1 Spacing Issues:
![image](https://github.com/BrooklinJazz/elixir_newbie_website_v2/assets/14877564/ab6e4ba2-b105-4892-93d5-c1c327fa20d6)

CSS flashing issues:
1. go to any page other than the blog show page.
2. refresh the current page to make sure the blog show page has not already rendered.
3. go to the blog show page through the navigation
4. briefly observe the following UI while some CSS appears to load in
![image](https://github.com/BrooklinJazz/elixir_newbie_website_v2/assets/14877564/b99536e7-554b-47ed-b88d-a3f40b758dfa)

BrooklinJazz commented:
Looks like the only work left to do is some CSS. Great job.

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+  ## alias ElixirNewbie.Router.Helpers, as: Routes ##
+
+
+  def render(assigns) do ## BigSpaces Wonders... about embedding the  {/if}  <figure class="makeup-blog"> as in v1
```
We could only use that syntax because we used Surface UI in the old project. Now we use embedded Elixir.

```elixir
<%= if condition do %>
some html
<% else %>
some other html
<% end %>

<% for n <- list do %>
some html
<% end %>
```

BigSpaces replied:
Thanks for clarifying!

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+
+
+      <section class="mx-auto w-2/3 pb-24">
+        <h1 class="mb-10 block rounded-lg text-4xl hover:opacity-75"><%= raw @blog.title%></h1>
```
is the blog title html? I don't think we need `raw` here.

BigSpaces replied:
Corrected!

BrooklinJazz commented on lib/elixir_newbie_web/live/blog_show.ex:
```diff
@@ -0,0 +1,62 @@
+        </a>
+        {/if} --%>
+        <div class="markdown">
+          <%= raw @blog.body%>
```
I'm being nitpicky, but formatting this stuff is important to make sure you are being "clean" otherwise code looks like a messy room with clothing everywhere.

```elixir
<%= raw @blog.body%>
should be
<%= raw @blog.body %>
```

It's missing a space. This might be best handled by a formatter, but as far as I'm aware there isn't great support for formatting HEEx inside of a `render` function.

BigSpaces replied:
Thanks. Nitpicky is appreciated. To me it is a marker of elegance and follow-through.

BigSpaces commented:
Work In Progress

Fixed:
- CSS full width on blog images
- h1 and h2 spacing issues in .markdown
- Removed commented code, save for router.ex line 22. Still troubleshooting.
- Formatted code

Not fixed:
- CSS flashing issues. Do not know how to troubleshoot that one.


Still to be implemented:

```
<.responsive_layout class="mt-12" scroll_id="all_blogs" cols={3} spacing="full">
 <%= for blog <- @blogs do %>
  <.blog_card blog={blog} />
 <% end %>
</.responsive_layout>
```

`@blogs` is not found. Need troubleshooting.
